File: scripts/db/V002_AddCartIndexes.js
// ============================================================
// V002 - Add Cart Indexes to carts collection
// DB Team standalone script
// Usage: mongosh "mongodb://user:pass@host:27017" --file V002_AddCartIndexes.js
// ============================================================

use('cart');

print('=== V002_AddCartIndexes START ===');

// Check if already executed
var existing = db.mongockChangeLog.findOne({
    changeId: 'V002_AddCartIndexes',
    state: 'EXECUTED'
});

if (existing) {
    print('  ⏭ V002 already executed at ' + existing.executedAt + ' - skipping');
    quit(0);
}

// 1. Unique index on user_id (one cart per user)
db.carts.createIndex(
    { user_id: 1 },
    { unique: true, name: 'idx_user_id_unique', background: true }
);
print('  ✓ idx_user_id_unique created (unique)');

// 2. Index on updated_at for sync queries
db.carts.createIndex(
    { updated_at: -1 },
    { name: 'idx_updated_at', background: true }
);
print('  ✓ idx_updated_at created');

// 3. Sparse index on synced_at (find un-synced carts)
db.carts.createIndex(
    { synced_at: 1 },
    { name: 'idx_synced_at', sparse: true, background: true }
);
print('  ✓ idx_synced_at created (sparse)');

// 4. Compound index: user_id + updated_at
db.carts.createIndex(
    { user_id: 1, updated_at: -1 },
    { name: 'idx_user_updated', background: true }
);
print('  ✓ idx_user_updated compound index created');

// Verify
var indexes = db.carts.getIndexes();
print('  Total indexes on carts: ' + indexes.length);
indexes.forEach(idx => print('    - ' + idx.name));

// Record in Mongock change log
db.mongockChangeLog.updateOne(
    { changeId: 'V002_AddCartIndexes' },
    {
        $set: {
            changeId:   'V002_AddCartIndexes',
            author:     'emart-db-team',
            state:      'EXECUTED',
            executedAt: new Date(),
            order:      '002',
            source:     'manual-script'
        }
    },
    { upsert: true }
);

print('=== V002_AddCartIndexes DONE ===');

File: cart-service/internal/migration/v002_add_cart_indexes.go
package migration

import (
	"context"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// V002AddCartIndexes adds required indexes to the carts collection
type V002AddCartIndexes struct{}
func NewV002AddCartIndexes() *V002AddCartIndexes { return &V002AddCartIndexes{} }
func (m *V002AddCartIndexes) ID() string     { return "V002_AddCartIndexes" }
func (m *V002AddCartIndexes) Order() string  { return "002" }
func (m *V002AddCartIndexes) Author() string { return "emart-db-team" }

func (m *V002AddCartIndexes) Execute(ctx context.Context, db *mongo.Database) error {
	coll := db.Collection("carts")
	indexes := []mongo.IndexModel{
		// Unique index on user_id - one cart per user
		{
			Keys:    bson.D{{Key: "user_id", Value: 1}},
			Options: options.Index().SetUnique(true).SetName("idx_user_id_unique"),
		},
		// Index on updated_at for sync queries (find recently updated)
		{
			Keys:    bson.D{{Key: "updated_at", Value: -1}},
			Options: options.Index().SetName("idx_updated_at"),
		},
		// Index on synced_at for finding unsynced carts
		{
			Keys:    bson.D{{Key: "synced_at", Value: 1}},
			Options: options.Index().SetSparse(true).SetName("idx_synced_at"),
		},
		// Compound index: user_id + updated_at for efficient queries
		{
			Keys:    bson.D{{Key: "user_id", Value: 1}, {Key: "updated_at", Value: -1}},
			Options: options.Index().SetName("idx_user_updated"),
		},
	}

	_, err := coll.Indexes().CreateMany(ctx, indexes)
	return err
}

func (m *V002AddCartIndexes) Rollback(ctx context.Context, db *mongo.Database) error {
	coll := db.Collection("carts")
	for _, name := range []string{"idx_user_id_unique", "idx_updated_at", "idx_synced_at", "idx_user_updated"} {
		coll.Indexes().DropOne(ctx, name)
	}
	return nil
}
